src/NServiceBus.Serverless/TransportWrapper/PipelineInvoker.cs

```diff
@@ -30,45 +30,36 @@ Task IPushMessages.Stop()
             return Task.CompletedTask;
         }
 
-        public async Task PushMessage(MessageContext messageContext)
+        public async Task PushFailedMessage(ErrorContext errorContext)
         {
-            var processingAttempts = 0;
+            var context = new ErrorContext(
+                errorContext.Exception,
+                new Dictionary<string, string>(errorContext.Message.Headers),
+                errorContext.Message.MessageId,
+                errorContext.Message.Body,
+                new TransportTransaction(),
+                errorContext.ImmediateProcessingFailures);
 
-            do
+            var errorHandling = await onError(context).ConfigureAwait(false);
+            if (errorHandling != ErrorHandleResult.Handled)
             {
-                try
-                {
-                    processingAttempts++;
-                    await onMessage.Invoke(new MessageContext(
-                            messageContext.MessageId,
-                            new Dictionary<string, string>(messageContext.Headers),
-                            messageContext.Body,
-                            messageContext.TransportTransaction,
-                            messageContext.ReceiveCancellationTokenSource,
-                            messageContext.Extensions))
-                        .ConfigureAwait(false);
-                    return;
-                }
-                catch (Exception exception)
-                {
-                    var errorContext = new ErrorContext(
-                        exception,
-                        new Dictionary<string, string>(messageContext.Headers),
-                        messageContext.MessageId,
-                        messageContext.Body,
-                        new TransportTransaction(),
-                        processingAttempts);
+                //handled means the message has been sent to the error queue
+                //otherwise it returns ErrorhandleResult.RetryRequired for immediate retries
+                //or throws an exception when not sending to the error queue
+                throw new Exception("Retry required");
```

Comment:
should we throw the original exception here, so that the caller will get the actual error instead of this message? Otherwise you'd have to use nested try/catch blocks and do some more weird stuff to return the original exception back to the user code?